feat: terraform provider support for login rules

nklaassen opened the pull request:
This PR adds support for [Login Rule](https://goteleport.com/docs/access-controls/login-rules/) resources to the Terraform provider.

The complicating factor here has been that the `LoginRule` type is defined outside of the "legacy" types package, in it's own new proto package [here](https://github.com/gravitational/teleport/blob/master/api/proto/teleport/loginrule/v1/loginrule.proto). This is the first type with the new proto layout to add Terraform support. For context, we are slowly moving away from the now-"legacy" types package because protoc-gen-gogo has been [deprecated](https://github.com/gogo/protobuf). New types are generated with [protoc-gen-go](https://protobuf.dev/reference/go/go-generated/).

Luckily, our [protoc-gen-terraform](https://github.com/gravitational/protoc-gen-terraform) plugin still ~(almost)~ works with the new proto layout. I have a small PR to fix an issue where it was not picking up the right import path https://github.com/gravitational/protoc-gen-terraform/pull/26 (merged).

The large diff here is mostly due to generated files, review the commits individually to avoid those.

This PR updates:
- `terraform/Makefile`, to generate Terraform schemas for the new loginrule package
- `_gen/main.go` to generate the Terraform plugin code for the new type
- the templates under `terraform/_gen/` to work around some type differences
- `terraform/provider/provider.go` to expose the new resource and date source types
- generated files in `terraform/provider/` are slightly different due to template updates

New files include:
- `terraform/protoc-gen-terraform-loginrule.yaml` which is copied/modified from `terraform/protoc-gen-terraform-teleport.yaml`
- files under `terraform/test/` for testing the provider with login rules
- new generated files under `terraform/provider/`

Check out `terraform/test/fixtures/login_rule_0_create.tf` for an example of the syntax for configuring a login rule with Terraform.

nklaassen reviewed (commented):
Note to reviewers: CI tests are currently failing because the teleport server being tested against is on v12.0.1 (defined in TELEPORT_GET_VERSION), which is missing [this](https://github.com/gravitational/teleport.e/pull/936) fix. I I'm going to see about updating the version to v12.0.5 once it is released

marcoandredinis commented:
> Note to reviewers: CI tests are currently failing because the teleport server being tested against is on v12.0.1 (defined in TELEPORT_GET_VERSION), which is missing [this](https://github.com/gravitational/teleport.e/pull/936) fix. I I'm going to see about updating the version to v12.0.5 once it is released

Feel free to rebase when this one gets merged
https://github.com/gravitational/teleport-plugins/pull/780

nklaassen commented:
> > Note to reviewers: CI tests are currently failing because the teleport server being tested against is on v12.0.1 (defined in TELEPORT_GET_VERSION), which is missing [this](https://github.com/gravitational/teleport.e/pull/936) fix. I I'm going to see about updating the version to v12.0.5 once it is released
> 
> Feel free to rebase when this one gets merged #780

I will rebase, but since 12.0.5 has test issues and that PR went back to 12.0.4, I'll probably need to wait for 12.0.6 for this

nklaassen commented:
@marcoandredinis @rosstimothy this should be ready for review now that ci is passing

marcoandredinis commented:
Can you please add an example to `terraform/examples/`?
The example role is missing the new resource kind (`terraform/examples/terraform.yaml`).

marcoandredinis commented on terraform/_gen/main.go:
```diff
@@ -231,6 +262,25 @@ var (
+		HasStaticID:       false,
+		ProtoPackage:      "loginrulev1",
+		ProtoPackagePath:  "github.com/gravitational/teleport/api/gen/proto/go/teleport/loginrule/v1",
+		SchemaPackage:     "schemav1",
```
:nail_care: 

```suggestion
		SchemaPackage:     "loginrulev1",
```

nklaassen replied:
That's already taken by `ProtoPackage`

marcoandredinis commented:
You get extra points if you add login rules resource here :joy: 
https://goteleport.com/docs/reference/terraform-provider/

nklaassen commented on terraform/test/loginrule_test.go:
```diff
@@ -0,0 +1,136 @@
+				ResourceName:  name,
+				ImportState:   true,
+				ImportStateId: id,
+				ImportStateCheck: func(state []*terraform.InstanceState) error {
```
there are not really any default values to check in login rules

nklaassen commented:
Hey @marcoandredinis I added login rule permissions to the example role, and added an example login rule resource in the most recent commits. I will be adding docs in a following pr to the teleport repo

marcoandredinis commented:
This PR should fix the current issue with `Plugin Tests (Go)` check
https://github.com/gravitational/teleport-plugins/pull/791
